File: 백준/Gold/24424. エゴイ展 （EGOI Exhibition）/エゴイ展 （EGOI Exhibition）.py
import sys
input=sys.stdin.readline

N=int(input())
M=int(input())

tree_size=M*2
tree=[0]*tree_size

def update(target,val):
    target+=M
    tree[target]=val
    while 1<target:
        target//=2
        tree[target]=max(tree[target*2],tree[target*2+1])

def get_val(left,right):
    left+=M
    right+=M
    ret=0
    while left<right:
        if left%2:
            ret=max(ret,tree[left])
            left+=1
        if right%2:
            right-=1
            ret=max(ret,tree[right])
        left//=2
        right//=2
    return ret

for _ in range(N):
    a,v=map(int,input().split())
    dp_a=get_val(a-1,a)
    update(a-1,max(v+max(get_val(0,a-1),get_val(a,M)),dp_a))
print(get_val(0,M))

File: 백준/Gold/24424. エゴイ展 （EGOI Exhibition）/エゴイ展 （EGOI Exhibition）.cc
#include <iostream>

using namespace std;

int N, M, a, v, dp_a;
int tree[600000] = {};


void update(int start, int end, int target, int val, int idx = 1) {
	if (end - start == 1) {
		tree[idx] = val;
		return;
	}
	int mid = (start + end) / 2;
	if (target < mid)
		update(start, mid, target, val, idx * 2);
	else
		update(mid, end, target, val, idx * 2 + 1);
	tree[idx] = max(tree[idx * 2], tree[idx * 2 + 1]);
}

int get_val(int start, int end, int left, int right, int idx = 1) {
	if (end <= left || right <= start || start == end)
		return 0;
	if (left <= start && end <= right)
		return tree[idx];
	int mid = (start + end) / 2;
	return max(get_val(start, mid, left, right, idx * 2), get_val(mid, end, left, right, idx * 2 + 1));
}

int main() {
	cin >> N >> M;
	for (int i = 0; i < N; i++) {
		cin >> a >> v;
		dp_a = get_val(0, M, a - 1, a);
		update(0, M, a - 1, max(v + max(get_val(0, M, 0, a - 1), get_val(0, M, a, M)), dp_a));
	}
	cout << get_val(0, M, 0, M);
}